reference/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,354 @@
+    ValueType* const t = s_hat + num_rows * num_rhs;
+    ValueType* const prec_work = t + num_rows * num_rhs;
+    ValueType rho_old[max_num_rhs];
+    ValueType rho_new[max_num_rhs];
+    ValueType omega[max_num_rhs];
+    ValueType alpha[max_num_rhs];
+    ValueType temp[max_num_rhs];
+    real_type norms_rhs[max_num_rhs];
+    real_type norms_res[max_num_rhs];
+
+    const auto A_entry = gko::batch::matrix::extract_batch_item(
+        gko::batch::matrix::to_const(a), batch_item_id);
+    const gko::batch::multi_vector::batch_item<const ValueType> b_entry =
+        gko::batch::extract_batch_item(gko::batch::to_const(b), batch_item_id);
+    const gko::batch::multi_vector::batch_item<ValueType> x_entry =
+        gko::batch::extract_batch_item(x, batch_item_id);
+
+    const gko::batch::multi_vector::batch_item<ValueType> r_entry{
+        r, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> r_hat_entry{
+        r_hat, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> p_entry{
+        p, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> p_hat_entry{
+        p_hat, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> v_entry{
+        v, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> s_entry{
+        s, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> s_hat_entry{
+        s_hat, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> t_entry{
+        t, num_rhs, num_rows, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> rho_old_entry{
+        rho_old, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> rho_new_entry{
+        rho_new, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> omega_entry{
+        omega, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> alpha_entry{
+        alpha, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<ValueType> temp_entry{
+        temp, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<real_type> rhs_norms_entry{
+        norms_rhs, num_rhs, 1, num_rhs};
+    const gko::batch::multi_vector::batch_item<real_type> res_norms_entry{
+        norms_res, num_rhs, 1, num_rhs};
+
+    // generate preconditioner
+    prec.generate(batch_item_id, A_entry, prec_work);
+
+    // initialization
+    // rho_old = 1, omega = 1, alpha = 1
+    // compute b norms
+    // r = b - A*x
+    // compute residual norms
+    // r_hat = r
+    // p = 0
+    // p_hat = 0
+    // v = 0
+    initialize(A_entry, b_entry, gko::batch::to_const(x_entry), rho_old_entry,
+               omega_entry, alpha_entry, r_entry, r_hat_entry, p_entry,
+               p_hat_entry, v_entry, rhs_norms_entry, res_norms_entry);
```

Comment:
the function call is slightly different from the core/solver/bicgstab. Is there any benefit merge b-Ax and r_hat = r to initialize? keeping them similar to core might be easier for reviewing

Reply:
I draw back my comment because the other kernel can put the dot together unlike the core already